experimental/sdk/webpubsub-socketio-extension/src/EIO/components/web-pubsub-transport.ts

```diff
@@ -1,28 +1,114 @@
+ * Similar with `Polling`, this transport always does batch send (send mulitple packets at once) and batch receive.
+ * Reference: https://github.com/socketio/engine.io/tree/6.4.2/lib/transports
  **/
 export class WebPubSubTransport extends Transport {
+  public clientConnectionContext: ClientConnectionContext;
+  /**
+   * Queue for storing packets to be sent. This design is to ensure packets are sent in order with send action in async manner.
+   */
+  public _queue: Packet[] = [];
+
+  // Reference: https://github.com/socketio/engine.io-parser/blob/5.0.7/lib/encodePacket.ts#L3
+  private _encodePacketAsync: (packet: Packet, supportsBinary: boolean) => Promise<RawData>;
+  // Reference: https://github.com/socketio/engine.io-parser/blob/5.0.7/lib/index.ts#L7
+  private _encodePayloadAsync: (packets: Packet[]) => Promise<string>;
+
+  /**
+   * Indicate whether any packet have been actually sent through AWPS or not.
+   */
+  private _onceSent: boolean;
+
+  constructor(req: unknown) {
+    debug("constructor");
+    super(req);
+    this.clientConnectionContext = req[WEBPUBSUB_CLIENT_CONNECTION_FILED_NAME];
+    this._onceSent = false;
+
+    // `Socket` places packets in its own buffer if `writable` == false. Otherwise, it calls `send` with buffer directly.
+    // Reference: https://github.com/socketio/engine.io/blob/6.4.2/lib/socket.ts#L510
+    this.writable = true;
+
+    this._encodePacketAsync = toAsync<string>(this.parser.encodePacket);
+    this._encodePayloadAsync = toAsync<string>(this.parser.encodePayload);
+  }
+
   public override supportsFraming = (): boolean => false;
 
   public override name = (): string => WEBPUBSUB_TRANSPORT_NAME;
 
-  public override async send(_packets: Packet[]): Promise<void> {
-    return Promise.resolve();
+  /**
+   * sends an array of `Packet` to the client.
+   * @param packets - An array of `Packet` to send
+   */
+  public override async send(packets: Packet[]): Promise<void> {
+    debug(`send packets, number = ${packets.length}`);
+    this.writable = false;
+
+    if (packets.length > 0 && !this._onceSent) {
+      const firstPacket = packets.shift();
+      if (firstPacket.type === "open") {
+        const payload = await this._encodePacketAsync(firstPacket, false);
+        debug(`first packet is 'open' packet, payload = ${payload}`);
+        this.clientConnectionContext.onAcceptEioConnection(payload.substring(1));
+        this._onceSent = true;
+      } else {
+        const errorMessage = `First packet must be 'open' packet, but got packet type = ${firstPacket.type}.`;
+        debug(errorMessage);
+        this.clientConnectionContext.onRefuseEioConnection(errorMessage);
+      }
+    }
+
+    if (packets.length > 0) {
+      // Clone queue and clear it immediately in case of `webPubSend` throws a exception without clearing queue.
+      const queue = packets.slice();
```

Comment:
`packets` is a local var, no need to copy?

Reply:
my bad. And I found the original `packets` have already been cleared in `socket.flush()` (where calls this `send` method). This part is no need.